keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java

```diff
@@ -833,6 +830,70 @@ public DoubleTensor concat(int dimension, DoubleTensor... those) {
         return new Nd4jDoubleTensor(concat);
     }
 
+    @Override
+    public DoubleTensor slice(int dimension, int index) {
+        INDArray dup = tensor.dup();
+        INDArray slice = dup.slice(index, dimension);
+        return new Nd4jDoubleTensor(slice);
+    }
+
+    /**
+     * @param dimension      the dimension to slice on
+     * @param splitAtIndices the indices that the dimension to slice on should be slice on
+     * @return pieces of the tensor slice in the order specified by splitAtIndices. To get
+     * pieces that encompasses the entire tensor, the last index in the splitAtIndices must
+     * be the length of the dimension being slice on.
+     * <p>
+     * e.g A =
+     * [
+     * 1, 2, 3, 4, 5, 6
+     * 7, 8, 9, 1, 2, 3
+     * ]
+     * <p>
+     * A.slice(0, [1]) gives List([1, 2, 3, 4, 5, 6])
+     * A.slice(0, [1, 2]) gives List([1, 2, 3, 4, 5, 6], [7, 8, 9, 1, 2, 3]
+     * <p>
+     * A.slice(1, [1, 3, 6]) gives
+     * List(
+     * [1, [2, 3  , [4, 5, 6,
+     * 7]  8, 9]    1, 2, 3]
+     * )
+     */
+    @Override
+    public List<DoubleTensor> split(int dimension, int[] splitAtIndices) {
+
+        int[] shape = getShape();
+        if (dimension < 0 || dimension >= shape.length) {
+            throw new IllegalArgumentException("Invalid dimension to split on " + dimension);
+        }
+
+        Nd4j.getCompressor().autoDecompress(tensor);
+
+        List<DoubleTensor> splits = new ArrayList<>();
+        int previousSplitIndex = 0;
+        for (int i = 0; i < splitAtIndices.length; i++) {
```

Comment:
We could use a foreach loop

Reply:
I thought it was easier to think about as iterating through the number of splits rather than iterating over the places to split.